src/midi/fluid_midi.c

```diff
@@ -2232,23 +2243,142 @@ int fluid_player_set_loop(fluid_player_t *player, int loop)
 }
 
 /**
- * Set the tempo of a MIDI player.
+ * update the MIDI player internal deltatime dependant of actual tempo.
  * @param player MIDI player instance
- * @param tempo Tempo to set playback speed to (in microseconds per quarter note, as per MIDI file spec)
- * @return Always returns #FLUID_OK
- * @note Tempo change events contained in the MIDI file can override the specified tempo at any time!
  */
-int fluid_player_set_midi_tempo(fluid_player_t *player, int tempo)
+static void fluid_player_update_tempo(fluid_player_t *player)
 {
-    player->miditempo = tempo;
-    player->deltatime = (double) tempo / player->division / 1000.0; /* in milliseconds */
+    int tempo; /* tempo in micro seconds by quarter note */
+    float deltatime;
+
+    if(fluid_atomic_int_get(&player->sync_mode))
+    {
+        /* take internal tempo from MIDI file */
+        tempo = fluid_atomic_int_get(&player->miditempo);
+    }
+    else
+    {
+        /* take  external tempo */
+        tempo = fluid_atomic_int_get(&player->exttempo);
+    }
+
+    /* compute deltattime (in ms) from current tempo and apply tempo multiplier */
+    deltatime = (float)tempo / (float)player->division / (float)1000.0;
+    deltatime /= fluid_atomic_float_get(&player->multempo); /* multiply tempo */
+    fluid_atomic_float_set(&player->deltatime, deltatime);
+
     player->start_msec = player->cur_msec;
     player->start_ticks = player->cur_ticks;
 
     FLUID_LOG(FLUID_DBG,
               "tempo=%d, tick time=%f msec, cur time=%d msec, cur tick=%d",
               tempo, player->deltatime, player->cur_msec, player->cur_ticks);
 
+}
+
+/**
+ * Set the tempo of a MIDI player.
+ * The player can be controlled by internal tempo comming from MIDI file tempo
+ * change or controlled by external tempo expressed in BPM or in micro seconds
+ * per quarter note.
+ *
+ * @param player MIDI player instance. Must be a valid pointer.
+ * @param tempo_type Must a be value of #fluid_player_tempo_type and indicates the
+ *  meaning of tempo value and how the player will be controlled, see below.
+ * @param tempo Tempo value or multiplier.
+ * 
+ *  #FLUID_PLAYER_TEMPO_DEFAULT, the player will be controlled by internal MIDI
+ *  file tempo changes. The @c tempo parameter will be ignored.
```

Comment:
Could you precise that if no tempo is given in the midi file, a default value of 120 bpm will be used?